_posts/2024-10-08-Internal-functioning-of-sorting-in-hybrid-query.md

```diff
@@ -0,0 +1,122 @@
+
+The following diagram illustrates the flow of the query phase and creating the final shard result based on sorting criteria.
+
+![hybrid-query-with-sorting](/assets/media/blog-images/2024-10-08-internal-functioning-of-sorting-in-hybrid-search/Hybrid-query-with-sorting.png){:class="img-centered"}
+
+
+## Detailed overview of sorting in a hybrid query 
+
+Building on the high-level logic discussed earlier, let's now dive into the detailed process of sorting in a hybrid query. This can be broken down into two key steps:
+
+1. Sorting individual subquery results at the shard level.
+2. Merging the multiple subquery results based on the sorting criteria at the coordinator node.
+
+Figure 1.2 illustrates these steps in detail. 
+
+![sorting-hld](/assets/media/blog-images/2024-10-08-internal-functioning-of-sorting-in-hybrid-search/Sorting-hld.png){:class="img-centered"}
+
+The coordinator node sends requests to data nodes to retrieve query results from the shards. Internally, during the query phase, the following process takes place:
+
+- A collector runs on each shard to gather results for each subquery. As results are collected, they are inserted into a priority queue that is ordered based on the sorting criteria.
+- At the end of the query phase, the subquery results are popped from the priority queue in sorted order.
+
+Once the coordinator node receives responses from all the data nodes, the normalization process begins. This process assigns a score to each subquery result and merges them into a final sorted list according to the specified sorting criteria. After normalization, the results move to the fetch phase, where the document IDs are used to retrieve the full document content from the shards. The final search response is then sent to the user.
+
+## Sorting limitations
+
+Because of the complexity involved in generating search results, sorting in hybrid queries has the following limitations:
+
+1. **Sorting by multiple fields cannot include `_score`**: You cannot sort by multiple fields if one of the fields is `_score` because subquery results can only be combined based on either the sort field or `_score`, but not both. Thus, a query similar to the following will return an error:
+  ```json
+  {
+    "sort":[
+      {
+        "foo":{
+          "order":"desc"
+        }
+      },
+      {
+        "_score":{
+          "order":"asc"
+      }       
+    ]
+  }
+  ```
+
+2. **Sorting is incompatible with `track_scores`**: When you specify `track_scores=true` in the search request, OpenSearch calculates scores during the fetch phase. However, this is impossible in hybrid queries. In a hybrid query, scores are initially calculated during the query phase and then normalized by the normalization processor. If `track_scores` is enabled, it triggers a recalculation of scores during sorting, which can produce incorrect results because the recalculated scores won’t reflect the normalization performed earlier in the process.
```

Comment:
```suggestion
2. **Sorting is incompatible with `track_scores`**: When you specify `track_scores=true` in the search request, OpenSearch calculates scores during the fetch phase. However, this is impossible in hybrid queries. In a hybrid query, scores are initially calculated during the query phase and then normalized by the normalization processor. If `track_scores` is enabled, it triggers a recalculation of scores during sorting, which can produce incorrect results because the recalculated scores won't reflect the normalization performed earlier in the process.
```